File: leetcode/pythonSolutions/minimum-number-of-arrows-to-burst-balloons.py

class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        
        
        s_list = sorted(points,key=lambda x:x[0])
        balloons = len(s_list)
        i = 1
        r = s_list[0][1]
        l = s_list[0][0]
        # l = s_list[j][0]
        print(s_list)

        while i<len(s_list):
            
            if r<s_list[i][0]:
                l = s_list[i][0]
                r = s_list[i][1]
                i+=1
                continue

            l = max(l,s_list[i][0])
            r = min(r,s_list[i][1])
            i+=1
            balloons-=1


        return balloons

"""
if you want to calculate the number arrows directly, use the below code
        if not points: return 0
        points.sort()
        prev = points[0]
        total = 1
        for s,e in points[1:]:
            if s>prev[1]:
                total += 1
                prev = [s,e]
            else:
                prev[1] = min(prev[1],e)

        return total
"""

File: leetcode/C++Solutions/minimum-number-of-arrows-to-burst-balloons.cpp

#include <vector>
#include <algorithm>

class Solution {
public:
    int findMinArrowShots(std::vector<std::vector<int>>& points) {
        // Sort intervals based on start points
        std::sort(points.begin(), points.end(), [](const std::vector<int>& a, const std::vector<int>& b) {
            return a[0] < b[0];
        });

        int balloons = points.size();
        int i = 1;
        int r = points[0][1];
        int l = points[0][0];

        while (i < points.size()) {
            if (r < points[i][0]) {
                l = points[i][0];
                r = points[i][1];
                i++;
                continue;
            }

            l = std::max(l, points[i][0]);
            r = std::min(r, points[i][1]);
            i++;
            balloons--;
        }

        return balloons;
    }
};

/* 
 *
 * I think the below code is another way of writing the above solution but little bit more efficiently
#include <vector>
#include <algorithm>

class Solution {
public:
    int findMinArrowShots(std::vector<std::vector<int>>& points) {
        if (points.empty()) return 0;

        int arrows = 1;
        std::sort(points.begin(), points.end(), [](const std::vector<int>& a, const std::vector<int>& b) {
            return a[1] < b[1];
        });

        int end = points[0][1];

        for (int i = 1; i < points.size(); ++i) {
            if (points[i][0] > end) {
                arrows++;
                end = points[i][1];
            }
        }

        return arrows;
    }
}; */
